### ISSUE TITLE
Cannot navigate in the app using the deep link if the app is already open

### ISSUE THREAD

**Issue by nishant09** [electron/electron#6820]
* Electron version: 0.34.1
* Operating system: 10.11

I've my electron app ready and open. When I click on custom deep link for e.g **my-app://navigate-to-this-resource**, I want to be able to handle this custom url,process it and navigate accordingly.

I could achieve this with the below code only **if the app is not opened yet**:

app.on("open-url", function(event, url) {
    (<any>global)[constants.macParams] = url;
});

How can I achieve this if the app is already open ? Should i do something in app.on("ready") event ? 
note : I can get the new paramaters if relaod the app from the help menu

Any help is appreciated

**Comment by MarshallOfSound** [comment]
@nishant09 If the app is already open that event is fired (I know because I implemented something similar last week).

However we no longer support Electron `0.34.1` so can you please update your Electron version to `1.3.3` and see if the issue persists.  If it does raise a new issue 👍

**Comment by nishant09** [comment]
@MarshallOfSound Thanks for your reply. I would appreciate if you could guide me little more here. How did you implemented it? Which event I need to listen to here. Would you be able to put the snippet ? I noticed after clicking my app gets the focus. however, It doesn't pass the url parameters. 

Yes, we are in the process of upgrading the electron version. will try to reproduce the issue once we have it.

**Comment by MarshallOfSound** [comment]
@nishant09  Literally my snippet looks like this

```js
app.on('open-url', (ev, data) => handle(data));

// handle
const handle = (data) => {
  let payload;
  try {
    payload = JSON.parse(unescape(data).replace(/^my-app-uri:\/\//g, '')).payload;
  } catch (err) {
    // ignore
  }
  if (payload) {
    handlePayload(payload);
  }
}
```

So it does *Just Work* 👍

**Comment by nishant09** [comment]
Appreciate. After digging more into this, I noticed the event gets fired. but the parameter I'm setting in the event(constants.macParams) is not getting passed to the renderer process. Have you seen this issue before ? How are you handling this situation in your code ? through IPC ?

**Comment by nishant09** [comment]
Forgive me some of the questions are silly. I'm pretty much new to the Electron and JS :)

**Comment by MarshallOfSound** [comment]
@nishant09 Can you post your code snippet and I'll take a look 👍

**Comment by nishant09** [comment]
@MarshallOfSound sure. here it is.
main.js
`/*!---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *----------------------------------------------------------*/

import "msint-identity-aad-electron";
import * as app from "app";
import * as BrowserWindow from "browser-window";
import * as dialog from "dialog";
import * as ipc from "ipc";
import * as menu from "menu";
import * as telemetryManager from "./renderer/telemetry/TelemetryManager";
import {Promise} from "q";

import constants = require("./Constants");

const packageInfo = require("../../package.json");

var mainWindow: GitHubElectron.BrowserWindow;

function showErrorMessage(title: string, message: string, buttons: string[], cancelId: number) {
    return Promise((resolve, reject) => {
        var focusedWindow = BrowserWindow.getFocusedWindow();
        dialog.showMessageBox(
            focusedWindow,
            {
                type: "error",
                message: message,
                title: title,
                cancelId: cancelId,
                buttons: buttons
            },
            resolve
        );
    });
}

function uncaughtJSErrorHandler(event: any, message, name, stack, source) {
    telemetryManager.sendError(
        "StorageExplorerUncaughtJSError",
        {
            message: message,
            name: name,
            stack: stack,
            source: source
        });

    return showErrorMessage(
        "Storage Explorer Error", // Localize
        `An unexpected error happened:\n\n${stack || message || name}`, // Localize
        ["Restart", "Close Application", "Ignore"], // Localize
        2).then((response) => {
            switch (response) {
                case 0:
                    // The user wants to restart, create the window again
                    createMainWindow();
                    break;
                case 1:
                    // The user wants to close the app.
                    process.exit(-1);
                    break;
                // Response 2, the user wants to ignore the error
                // and continue running.
            }
        });
};

function rendererProcessCrashHandler() {
    telemetryManager.sendError("StorageExplorerRendererProcessCrash");

    return showErrorMessage(
        "Storage Explorer has stopped working", // Localize
        "A problem caused Storage Explorer to stop working correctly. Storage Explorer can check for a solution and try to restart the program.", // Localize
        ["Restart", "Close Application"], // Localize
        1).then((response) => {
            switch (response) {
                case 0:
                    //  The user wants to restart, create the window again
                    createMainWindow();
                    break;
                default:
                    // The user wants to close the app.
                    process.exit(-1);
            }
        });
};

function mainProcessGPUCrashHandler() {
    telemetryManager.sendError("StorageExplorerMainProcessGPUCrash");

    return showErrorMessage(
        "Storage Explorer has stopped working", // Localize
        "A problem caused Storage Explorer to stop working correctly. Please close the program.", // Localize
        ["Close Application"],
        0).then(() => {
            process.exit(-1);
        });
};

function onMainWindowClosed() {
    mainWindow = null;
    app.quit();
};

function createMainWindow() {
    // Destroy the window if it was open already.
    if (mainWindow) {
        mainWindow.removeListener("closed", onMainWindowClosed);
        mainWindow.close();
        mainWindow = null;
    }

    // Create the window
    mainWindow = new BrowserWindow({
        title: packageInfo.displayName,
        width: 1000,
        height: 768,
        "min-width": 1000,
        "min-height": 768,
        icon: __dirname + "/icon.png" // Linux icon on launcher
    });

    // Enable this to show developer tools if there is an exception before the application is fully initialized
    // mainWindow.toggleDevTools();

    var preventBubbling = (e: Event) => {
        e.preventDefault();
        return false;
    };

    // Prevent the window to navigate to other pages.
    // i.e: When droping a file to the app.
    mainWindow.webContents.on("will-navigate", preventBubbling);

    // Remove the reference of the window to avoid memory leaks
    // when the app is closed.
    mainWindow.on("closed", onMainWindowClosed);

    // Handle process crashes unexpectly
    // Documentation: http://electron.atom.io/docs/all/#event-crashed
    mainWindow.webContents.on("crashed", rendererProcessCrashHandler);

    // Load the main page.
    mainWindow.loadUrl("file://" + __dirname + "/renderer/index.html");
}

// Handle gpu process crashes unexpectly
// Documentation: http://electron.atom.io/docs/all/#event-gpu-process-crashed
app.on("gpu-process-crashed", mainProcessGPUCrashHandler);

// To get custom url proocol in OSX, we need to listen to the open-url event 
// before the ready event since your app is registered to URL schemes
app.on("open-url", function (event, url) {
    (<any>global)[constants.macParams] = url;
});

// Handle uncaught exceptions from rederer processes
ipc.on(constants.UncaughtExceptionRendererEvent, uncaughtJSErrorHandler);

// Get cmd arguments if APP is started from deeplink.
(<any>global)[constants.sharedObjectName] = { args: process.argv };

app.on("ready", function () {
    // Set the AppId early once the app is ready.
    // This allowes the OS to identify the app instance and
    // not confuse it with another Electron app.
    // This needs to be executed right after the app is ready.
    app.setAppUserModelId("com.microsoft.StorageExplorer");
    
    // Prevent the default Menu to appear
    var menuConfig = menu.buildFromTemplate([]);
    menu.setApplicationMenu(menuConfig);

    createMainWindow();
});
`
constants.macParams is globally defined and I'm using it in my renderer process as shown below:
           var args: string[] = remote.getGlobal(constants.sharedObjectName).args;
            args.push(remote.getGlobal(constants.macParams));

**Comment by nishant09** [comment]
hope the code is in the readable state. This code works ok if the app is not launched yet.
But I want to make it work even when the app is already launched and click on the deep link. My renderer process should be invoked as soon as I click on the deep link and "open-url" gets triggered and Main process should pass the updated value of the contants.macParams to the renderer process

**Comment by MarshallOfSound** [comment]
> My renderer process should be invoked as soon as I click on the deep link and "open-url" gets triggered and Main process should pass the updated value of the contants.macParams to the renderer process

@nishant09 From a quick look through your code you aren't even trying to do that.  Setting a global isn't going to magically reload the renderer process and pass in the new global.  Things don't work like that 😕 

You would need to send the new value that you received in `open-url` into the existing renderer process with IPC

**Comment by nishant09** [comment]
@MarshallOfSound got it. Thanks for reviewing and your feedback. I will look into IPC now.

**Comment by nishant09** [comment]
could you name the portal where we can ask/discuss questions related to Electron ?
